src/function/window/window_rownumber_function.cpp

```diff
@@ -10,14 +81,100 @@ WindowRowNumberExecutor::WindowRowNumberExecutor(BoundWindowExpression &wexpr, C
+	return make_uniq<WindowRowNumberGlobalState>(*this, payload_count, partition_mask, order_mask);
+}
+
+unique_ptr<WindowExecutorLocalState>
+WindowRowNumberExecutor::GetLocalState(const WindowExecutorGlobalState &gstate) const {
+	return make_uniq<WindowRowNumberLocalState>(gstate.Cast<WindowRowNumberGlobalState>());
+}
+
 void WindowRowNumberExecutor::EvaluateInternal(WindowExecutorGlobalState &gstate, WindowExecutorLocalState &lstate,
                                                DataChunk &eval_chunk, Vector &result, idx_t count,
                                                idx_t row_idx) const {
-	auto &lbstate = lstate.Cast<WindowExecutorBoundsState>();
-	auto partition_begin = FlatVector::GetData<const idx_t>(lbstate.bounds.data[PARTITION_BEGIN]);
+	auto &grstate = gstate.Cast<WindowRowNumberGlobalState>();
+	auto &lrstate = lstate.Cast<WindowRowNumberLocalState>();
+	auto partition_begin = FlatVector::GetData<const idx_t>(lrstate.bounds.data[PARTITION_BEGIN]);
+	auto rdata = FlatVector::GetData<uint64_t>(result);
+
+	if (grstate.token_tree) {
+		auto partition_end = FlatVector::GetData<const idx_t>(lrstate.bounds.data[PARTITION_END]);
+		for (idx_t i = 0; i < count; ++i, ++row_idx) {
+			// Row numbers are unique ranks
+			rdata[i] = grstate.token_tree->Rank(partition_begin[i], partition_end[i], row_idx);
+		}
+		return;
+	}
+
+	for (idx_t i = 0; i < count; ++i, ++row_idx) {
+		rdata[i] = row_idx - partition_begin[i] + 1;
+	}
+}
+
+//===--------------------------------------------------------------------===//
+// WindowNtileExecutor
+//===--------------------------------------------------------------------===//
+WindowNtileExecutor::WindowNtileExecutor(BoundWindowExpression &wexpr, ClientContext &context,
+                                         WindowSharedExpressions &shared)
+    : WindowRowNumberExecutor(wexpr, context, shared) {
+
+	// NTILE has one argument
+	ntile_idx = shared.RegisterEvaluate(wexpr.children[0]);
+}
+
+void WindowNtileExecutor::EvaluateInternal(WindowExecutorGlobalState &gstate, WindowExecutorLocalState &lstate,
+                                           DataChunk &eval_chunk, Vector &result, idx_t count, idx_t row_idx) const {
+	auto &grstate = gstate.Cast<WindowRowNumberGlobalState>();
+	auto &lrstate = lstate.Cast<WindowRowNumberLocalState>();
+	auto partition_begin = FlatVector::GetData<const idx_t>(lrstate.bounds.data[PARTITION_BEGIN]);
+	auto partition_end = FlatVector::GetData<const idx_t>(lrstate.bounds.data[PARTITION_END]);
 	auto rdata = FlatVector::GetData<int64_t>(result);
+	WindowInputExpression ntile_col(eval_chunk, ntile_idx);
 	for (idx_t i = 0; i < count; ++i, ++row_idx) {
-		rdata[i] = NumericCast<int64_t>(row_idx - partition_begin[i] + 1);
+		if (ntile_col.CellIsNull(i)) {
+			FlatVector::SetNull(result, i, true);
+		} else {
+			auto n_param = ntile_col.GetCell<int64_t>(i);
+			if (n_param < 1) {
+				throw InvalidInputException("Argument for ntile must be greater than zero");
```

Comment:
Can this exception be caught before execution? Regardless, can we add a test that triggers this?

Reply:
Heh - this is code that Hannes wrote back in the day - I just moved it. But I can add a test.

As for catching it before execution, it is an expression, not a constant (which is rather nonsensical but there you are) so we may not even be able to tell. Consider `FLOOR(SIN(x / 1000))`...

Reply:
Actually, there is already a test: see `test/sql/window/test_ntile.test` at line 163

Reply:
That makes sense. Thanks